<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>9 Iterating over PyRanges | PyRanges documentation</title>
  <meta name="description" content="9 Iterating over PyRanges | PyRanges documentation" />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="9 Iterating over PyRanges | PyRanges documentation" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="9 Iterating over PyRanges | PyRanges documentation" />
  
  
  

<meta name="author" content="Endre Bakken Stovner" />


<meta name="date" content="2020-02-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="printing-pyranges.html">
<link rel="next" href="sorting-pyranges.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
<li class="chapter" data-level="32" data-path="fetching-gene-tracks-as-pyranges.html"><a href="fetching-gene-tracks-as-pyranges.html"><i class="fa fa-check"></i><b>32</b> Fetching gene tracks as PyRanges</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">PyRanges documentation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="iterating-over-pyranges" class="section level1">
<h1><span class="header-section-number">9</span> Iterating over PyRanges</h1>
<p>PyRanges can be iterated over by a simple for loop, and the data is guaranteed
to come back in natsorted order:</p>
<div class="sourceCode" id="cb105"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb105-1" data-line-number="1"><span class="im">import</span> pyranges <span class="im">as</span> pr</a>
<a class="sourceLine" id="cb105-2" data-line-number="2"></a>
<a class="sourceLine" id="cb105-3" data-line-number="3"><span class="im">import</span> pandas <span class="im">as</span> pd</a>
<a class="sourceLine" id="cb105-4" data-line-number="4"></a>
<a class="sourceLine" id="cb105-5" data-line-number="5">exons <span class="op">=</span> pr.data.exons()</a>
<a class="sourceLine" id="cb105-6" data-line-number="6">cpg <span class="op">=</span> pr.data.cpg()</a>
<a class="sourceLine" id="cb105-7" data-line-number="7"></a>
<a class="sourceLine" id="cb105-8" data-line-number="8"><span class="cf">for</span> k, df <span class="kw">in</span> cpg:</a>
<a class="sourceLine" id="cb105-9" data-line-number="9">    <span class="bu">print</span>(k)</a>
<a class="sourceLine" id="cb105-10" data-line-number="10">    <span class="bu">print</span>(df.head(<span class="dv">3</span>))</a></code></pre></div>
<pre><code>## chrX
##   Chromosome   Start     End  CpG
## 0       chrX   64181   64793   62
## 1       chrX   69133   70029  100
## 2       chrX  148685  149461   85
## chrY
##     Chromosome  Start    End  CpG
## 896       chrY  14181  14793   62
## 897       chrY  19133  20029  100
## 898       chrY  98685  99461   85</code></pre>
<p>If the data is unstranded, the key is just the chromosome names, but if the data
is stranded the key is a tuple of the chromosome and strand:</p>
<div class="sourceCode" id="cb107"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb107-1" data-line-number="1"><span class="cf">for</span> k, df <span class="kw">in</span> exons:</a>
<a class="sourceLine" id="cb107-2" data-line-number="2">    <span class="bu">print</span>(k)</a>
<a class="sourceLine" id="cb107-3" data-line-number="3">    <span class="bu">print</span>(df.head(<span class="dv">3</span>))</a></code></pre></div>
<pre><code>## (&#39;chrX&#39;, &#39;+&#39;)
##   Chromosome      Start        End                                 Name  Score Strand
## 0       chrX  135721701  135721963  NR_038462_exon_0_0_chrX_135721702_f      0      +
## 2       chrX  135574120  135574598  NM_001727_exon_2_0_chrX_135574121_f      0      +
## 3       chrX   47868945   47869126   NM_205856_exon_4_0_chrX_47868946_f      0      +
## (&#39;chrX&#39;, &#39;-&#39;)
##   Chromosome      Start        End                                    Name  Score Strand
## 1       chrX   49069126   49069255  NM_001256790_exon_15_0_chrX_49069127_r      0      -
## 4       chrX  154006958  154007628  NM_001166460_exon_0_0_chrX_154006959_r      0      -
## 6       chrX   52257919   52258046   NM_001097593_exon_1_0_chrX_52257920_r      0      -
## (&#39;chrY&#39;, &#39;+&#39;)
##    Chromosome     Start       End                                  Name  Score Strand
## 14       chrY   1693161   1693291  NM_001171039_exon_2_0_chrY_1693162_f      0      +
## 82       chrY   1357411   1357535  NM_001161531_exon_4_0_chrY_1357412_f      0      +
## 98       chrY  59233166  59233257    NR_033715_exon_4_0_chrY_59233167_f      0      +
## (&#39;chrY&#39;, &#39;-&#39;)
##   Chromosome     Start       End                                 Name  Score Strand
## 5       chrY   1481624   1481747    NM_004192_exon_1_0_chrY_1481625_r      0      -
## 8       chrY  15526614  15526673  NR_047607_exon_25_0_chrY_15526615_r      0      -
## 9       chrY  15591393  15592550  NR_047610_exon_27_0_chrY_15591394_r      0      -</code></pre>
<p>If you would like to iterate over the chromosomes in a stranded PyRange the idiom is</p>
<div class="sourceCode" id="cb109"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb109-1" data-line-number="1"><span class="cf">for</span> c <span class="kw">in</span> exons.chromosomes:</a>
<a class="sourceLine" id="cb109-2" data-line-number="2">    <span class="bu">print</span>(c)</a>
<a class="sourceLine" id="cb109-3" data-line-number="3">    df <span class="op">=</span> exons[c].df</a>
<a class="sourceLine" id="cb109-4" data-line-number="4">    <span class="bu">print</span>(df.head())</a></code></pre></div>
<pre><code>## chrX
##   Chromosome      Start        End                                   Name  Score Strand
## 0       chrX  135721701  135721963    NR_038462_exon_0_0_chrX_135721702_f      0      +
## 1       chrX  135574120  135574598    NM_001727_exon_2_0_chrX_135574121_f      0      +
## 2       chrX   47868945   47869126     NM_205856_exon_4_0_chrX_47868946_f      0      +
## 3       chrX   77294333   77294480    NM_000052_exon_17_0_chrX_77294334_f      0      +
## 4       chrX   91090459   91091043  NM_001168360_exon_0_0_chrX_91090460_f      0      +
## chrY
##   Chromosome     Start       End                                  Name  Score Strand
## 0       chrY   1693161   1693291  NM_001171039_exon_2_0_chrY_1693162_f      0      +
## 1       chrY   1357411   1357535  NM_001161531_exon_4_0_chrY_1357412_f      0      +
## 2       chrY  59233166  59233257    NR_033715_exon_4_0_chrY_59233167_f      0      +
## 3       chrY   1693161   1693291  NM_001171038_exon_2_0_chrY_1693162_f      0      +
## 4       chrY   1664276   1664425     NM_005088_exon_2_0_chrY_1664277_f      0      +</code></pre>
<p>but notice that we need the .df accessor, because subsetting a PyRange always
returns a PyRange.</p>
<p>There are three more ways to iterate over a pyrange, namely the keys, values and
items methods. These return a list, not a custom object like the python dict.</p>
<p>If you want to create a new PyRange while iterating over it, the idiom is</p>
<div class="sourceCode" id="cb111"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb111-1" data-line-number="1">d <span class="op">=</span> {}</a>
<a class="sourceLine" id="cb111-2" data-line-number="2"><span class="cf">for</span> k, df <span class="kw">in</span> exons:</a>
<a class="sourceLine" id="cb111-3" data-line-number="3">    <span class="bu">print</span>(k)</a>
<a class="sourceLine" id="cb111-4" data-line-number="4">    d[k] <span class="op">=</span> df.head(<span class="dv">3</span>)</a></code></pre></div>
<pre><code>## (&#39;chrX&#39;, &#39;+&#39;)
## (&#39;chrX&#39;, &#39;-&#39;)
## (&#39;chrY&#39;, &#39;+&#39;)
## (&#39;chrY&#39;, &#39;-&#39;)</code></pre>
<div class="sourceCode" id="cb113"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb113-1" data-line-number="1">new_gr <span class="op">=</span> pr.PyRanges(d)</a>
<a class="sourceLine" id="cb113-2" data-line-number="2"><span class="bu">print</span>(new_gr)</a></code></pre></div>
<pre><code>## +--------------+-----------+-----------+----------------------------------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name                                   | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)                               | (int64)   | (category)   |
## |--------------+-----------+-----------+----------------------------------------+-----------+--------------|
## | chrX         | 135721701 | 135721963 | NR_038462_exon_0_0_chrX_135721702_f    | 0         | +            |
## | chrX         | 135574120 | 135574598 | NM_001727_exon_2_0_chrX_135574121_f    | 0         | +            |
## | chrX         | 47868945  | 47869126  | NM_205856_exon_4_0_chrX_47868946_f     | 0         | +            |
## | chrX         | 49069126  | 49069255  | NM_001256790_exon_15_0_chrX_49069127_r | 0         | -            |
## | ...          | ...       | ...       | ...                                    | ...       | ...          |
## | chrY         | 59233166  | 59233257  | NR_033715_exon_4_0_chrY_59233167_f     | 0         | +            |
## | chrY         | 1481624   | 1481747   | NM_004192_exon_1_0_chrY_1481625_r      | 0         | -            |
## | chrY         | 15526614  | 15526673  | NR_047607_exon_25_0_chrY_15526615_r    | 0         | -            |
## | chrY         | 15591393  | 15592550  | NR_047610_exon_27_0_chrY_15591394_r    | 0         | -            |
## +--------------+-----------+-----------+----------------------------------------+-----------+--------------+
## Stranded PyRanges object has 12 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.</code></pre>
<p>Note that this is basically the same as using the apply method of the PyRange,
only that the for loops are never multithreaded.</p>
<p>To iterate over multiple pyranges you can use pr.itergrs or pr.iter. It iterates
over multiple pyranges at the same time, returning the dfs belonging to the same
Chromosome or Chromosome/Strand at each iteration. Missing entries in one or
more PyRanges returns empty dfs.</p>
<div class="sourceCode" id="cb115"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb115-1" data-line-number="1"><span class="im">import</span> pyranges <span class="im">as</span> pr</a>
<a class="sourceLine" id="cb115-2" data-line-number="2"></a>
<a class="sourceLine" id="cb115-3" data-line-number="3">l <span class="op">=</span> pr.random(<span class="dv">25</span>), pr.random(<span class="dv">25</span>), pr.random(<span class="dv">25</span>)</a>
<a class="sourceLine" id="cb115-4" data-line-number="4"></a>
<a class="sourceLine" id="cb115-5" data-line-number="5"><span class="cf">for</span> key, grs <span class="kw">in</span> pr.<span class="bu">iter</span>(l, keys<span class="op">=</span><span class="va">True</span>, strand<span class="op">=</span><span class="va">True</span>):</a>
<a class="sourceLine" id="cb115-6" data-line-number="6">    <span class="bu">print</span>(key)</a>
<a class="sourceLine" id="cb115-7" data-line-number="7">    <span class="bu">print</span>(grs)</a></code></pre></div>
<pre><code>## (&#39;chr20&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome    Start      End Strand
## 0      chr20  8138908  8139008      -,   Chromosome     Start       End Strand
## 1      chr20  28194163  28194263      -]
## (&#39;chr14&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr14  26953180  26953280      +,   Chromosome     Start       End Strand
## 0      chr14  96401150  96401250      +
## 1      chr14  96413256  96413356      +]
## (&#39;chr3&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0       chr3  124466989  124467089      -
## 1       chr3   13642007   13642107      -,   Chromosome      Start        End Strand
## 1       chr3  168306282  168306382      -]
## (&#39;chr21&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr21  31292285  31292385      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr2&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0       chr2  107773861  107773961      -
## 1       chr2   54686939   54687039      -]
## (&#39;chr7&#39;, &#39;-&#39;)
## [  Chromosome      Start        End Strand
## 0       chr7   17943711   17943811      -
## 1       chr7  157652274  157652374      -
## 2       chr7  133675538  133675638      -,   Chromosome      Start        End Strand
## 0       chr7  149972489  149972589      -,   Chromosome     Start       End Strand
## 0       chr7  79710047  79710147      -
## 1       chr7   4183867   4183967      -]
## (&#39;chr17&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 0      chr17  74245342  74245442      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr6&#39;, &#39;+&#39;)
## [  Chromosome      Start        End Strand
## 0       chr6  108762618  108762718      +,   Chromosome     Start       End Strand
## 1       chr6  18518627  18518727      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chrX&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0       chrX  131298446  131298546      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr19&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1      chr19  43209866  43209966      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr4&#39;, &#39;-&#39;)
## [  Chromosome      Start        End Strand
## 1       chr4  181566043  181566143      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0       chr4  22328930  22329030      -
## 1       chr4  15751892  15751992      -]
## (&#39;chr12&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1      chr12  29398471  29398571      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr17&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr17  59069253  59069353      +]
## (&#39;chr10&#39;, &#39;+&#39;)
## [  Chromosome    Start      End Strand
## 1      chr10  7492406  7492506      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr9&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 1       chr9  120171189  120171289      -
## 3       chr9   34301318   34301418      -]
## (&#39;chr19&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr19  16650294  16650394      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chrY&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 0       chrY  53202002  53202102      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr1&#39;, &#39;-&#39;)
## [  Chromosome      Start        End Strand
## 0       chr1  132458119  132458219      -
## 3       chr1   28499072   28499172      -
## 4       chr1  209314949  209315049      -
## 6       chr1  162373989  162374089      -,   Chromosome      Start        End Strand
## 0       chr1  212894110  212894210      -,   Chromosome      Start        End Strand
## 1       chr1  109330899  109330999      -
## 2       chr1  106028704  106028804      -]
## (&#39;chr5&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1       chr5  99644685  99644785      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr20&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1      chr20  37547541  37547641      +,   Chromosome     Start       End Strand
## 0      chr20  26351212  26351312      +]
## (&#39;chr13&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0      chr13  100866481  100866581      +]
## (&#39;chr3&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 2       chr3  93527177  93527277      +,   Chromosome     Start       End Strand
## 0       chr3  90489313  90489413      +]
## (&#39;chr22&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr22  16003098  16003198      -,   Chromosome     Start       End Strand
## 0      chr22  20543358  20543458      -]
## (&#39;chr8&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1       chr8  71433208  71433308      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr1&#39;, &#39;+&#39;)
## [  Chromosome      Start        End Strand
## 1       chr1   67273847   67273947      +
## 2       chr1   26272408   26272508      +
## 5       chr1  205340014  205340114      +
## 7       chr1  123702454  123702554      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0       chr1  126937764  126937864      +]
## (&#39;chr15&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0      chr15  91946763  91946863      +,   Chromosome     Start       End Strand
## 0      chr15  91172292  91172392      +,   Chromosome     Start       End Strand
## 0      chr15  19281885  19281985      +]
## (&#39;chr5&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 0       chr5  88501433  88501533      -,   Chromosome     Start       End Strand
## 0       chr5  29481603  29481703      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr16&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0      chr16  48448301  48448401      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chrX&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 0       chrX  88907424  88907524      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr4&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0       chr4  12754845  12754945      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr12&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr12  10029261  10029361      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr7&#39;, &#39;+&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 1       chr7  10239606  10239706      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr10&#39;, &#39;-&#39;)
## [  Chromosome      Start        End Strand
## 0      chr10  132015880  132015980      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0      chr10  54151437  54151537      -]
## (&#39;chr18&#39;, &#39;-&#39;)
## [  Chromosome     Start       End Strand
## 0      chr18  77921690  77921790      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [], Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr9&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0       chr9  12199321  12199421      +, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome      Start        End Strand
## 0       chr9  113342950  113343050      +
## 2       chr9   74764944   74765044      +]
## (&#39;chr8&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0       chr8  49096849  49096949      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]
## (&#39;chr11&#39;, &#39;+&#39;)
## [  Chromosome     Start       End Strand
## 0      chr11  84953305  84953405      +,   Chromosome     Start       End Strand
## 0      chr11  84352745  84352845      +,   Chromosome     Start       End Strand
## 0      chr11  10939480  10939580      +]
## (&#39;chr6&#39;, &#39;-&#39;)
## [Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: [],   Chromosome     Start       End Strand
## 0       chr6  94634172  94634272      -
## 2       chr6  20389068  20389168      -, Empty DataFrame
## Columns: [Chromosome, Start, End, Strand]
## Index: []]</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="printing-pyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="sorting-pyranges.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
