<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Printing PyRanges | PyRanges documentation</title>
  <meta name="description" content="8 Printing PyRanges | PyRanges documentation" />
  <meta name="generator" content="bookdown 0.12 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Printing PyRanges | PyRanges documentation" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Printing PyRanges | PyRanges documentation" />
  
  
  

<meta name="author" content="Endre Bakken Stovner" />


<meta name="date" content="2020-02-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="piping-pyranges.html">
<link rel="next" href="iterating-over-pyranges.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction to PyRanges</a></li>
<li class="chapter" data-level="2" data-path="loadingcreating-pyranges.html"><a href="loadingcreating-pyranges.html"><i class="fa fa-check"></i><b>2</b> Loading/Creating PyRanges</a></li>
<li class="chapter" data-level="3" data-path="writing-pyranges-to-disk.html"><a href="writing-pyranges-to-disk.html"><i class="fa fa-check"></i><b>3</b> Writing PyRanges to disk</a></li>
<li class="chapter" data-level="4" data-path="subsetting-pyranges.html"><a href="subsetting-pyranges.html"><i class="fa fa-check"></i><b>4</b> Subsetting PyRanges</a></li>
<li class="chapter" data-level="5" data-path="manipulating-the-data-in-pyranges.html"><a href="manipulating-the-data-in-pyranges.html"><i class="fa fa-check"></i><b>5</b> Manipulating the data in PyRanges</a></li>
<li class="chapter" data-level="6" data-path="concatenating-pyranges.html"><a href="concatenating-pyranges.html"><i class="fa fa-check"></i><b>6</b> Concatenating PyRanges</a></li>
<li class="chapter" data-level="7" data-path="piping-pyranges.html"><a href="piping-pyranges.html"><i class="fa fa-check"></i><b>7</b> Piping PyRanges</a></li>
<li class="chapter" data-level="8" data-path="printing-pyranges.html"><a href="printing-pyranges.html"><i class="fa fa-check"></i><b>8</b> Printing PyRanges</a></li>
<li class="chapter" data-level="9" data-path="iterating-over-pyranges.html"><a href="iterating-over-pyranges.html"><i class="fa fa-check"></i><b>9</b> Iterating over PyRanges</a></li>
<li class="chapter" data-level="10" data-path="sorting-pyranges.html"><a href="sorting-pyranges.html"><i class="fa fa-check"></i><b>10</b> Sorting PyRanges</a></li>
<li class="chapter" data-level="11" data-path="summarizing-pyranges.html"><a href="summarizing-pyranges.html"><i class="fa fa-check"></i><b>11</b> Summarizing PyRanges</a></li>
<li class="chapter" data-level="12" data-path="methods-for-manipulating-single-pyranges.html"><a href="methods-for-manipulating-single-pyranges.html"><i class="fa fa-check"></i><b>12</b> Methods for manipulating single PyRanges</a></li>
<li class="chapter" data-level="13" data-path="applying-custom-methods-to-pairs-of-pyranges.html"><a href="applying-custom-methods-to-pairs-of-pyranges.html"><i class="fa fa-check"></i><b>13</b> Applying custom methods to pairs of PyRanges</a></li>
<li class="chapter" data-level="14" data-path="intersecting-ranges.html"><a href="intersecting-ranges.html"><i class="fa fa-check"></i><b>14</b> Intersecting Ranges</a></li>
<li class="chapter" data-level="15" data-path="overlapping-ranges.html"><a href="overlapping-ranges.html"><i class="fa fa-check"></i><b>15</b> Overlapping Ranges</a></li>
<li class="chapter" data-level="16" data-path="joining-ranges.html"><a href="joining-ranges.html"><i class="fa fa-check"></i><b>16</b> Joining Ranges</a></li>
<li class="chapter" data-level="17" data-path="finding-the-closest-intervals.html"><a href="finding-the-closest-intervals.html"><i class="fa fa-check"></i><b>17</b> Finding the closest intervals</a></li>
<li class="chapter" data-level="18" data-path="finding-the-k-closest-intervals.html"><a href="finding-the-k-closest-intervals.html"><i class="fa fa-check"></i><b>18</b> Finding the k closest intervals</a></li>
<li class="chapter" data-level="19" data-path="create-count-matrix-from-multiple-pyranges.html"><a href="create-count-matrix-from-multiple-pyranges.html"><i class="fa fa-check"></i><b>19</b> Create count-matrix from multiple PyRanges</a></li>
<li class="chapter" data-level="20" data-path="statistics-colocalization-and-co-occurence-measures.html"><a href="statistics-colocalization-and-co-occurence-measures.html"><i class="fa fa-check"></i><b>20</b> Statistics: colocalization and co-occurence measures</a></li>
<li class="chapter" data-level="21" data-path="statistics-simes-method.html"><a href="statistics-simes-method.html"><i class="fa fa-check"></i><b>21</b> Statistics: Simes method</a></li>
<li class="chapter" data-level="22" data-path="statistics-compute-fisher-exact.html"><a href="statistics-compute-fisher-exact.html"><i class="fa fa-check"></i><b>22</b> Statistics: compute Fisher exact</a></li>
<li class="chapter" data-level="23" data-path="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><a href="statistics-computing-the-matthews-correlation-coeficcient-between-ranges.html"><i class="fa fa-check"></i><b>23</b> Statistics: computing the Matthews correlation coeficcient between ranges</a></li>
<li class="chapter" data-level="24" data-path="statistics-rowbased-operations.html"><a href="statistics-rowbased-operations.html"><i class="fa fa-check"></i><b>24</b> Statistics: rowbased operations</a></li>
<li class="chapter" data-level="25" data-path="turning-ranges-into-rles.html"><a href="turning-ranges-into-rles.html"><i class="fa fa-check"></i><b>25</b> Turning Ranges into RLEs</a></li>
<li class="chapter" data-level="26" data-path="an-introduction-to-rles.html"><a href="an-introduction-to-rles.html"><i class="fa fa-check"></i><b>26</b> An introduction to RLEs</a></li>
<li class="chapter" data-level="27" data-path="run-length-encoding-dicts.html"><a href="run-length-encoding-dicts.html"><i class="fa fa-check"></i><b>27</b> Run Length Encoding dicts</a></li>
<li class="chapter" data-level="28" data-path="subsetting-rles.html"><a href="subsetting-rles.html"><i class="fa fa-check"></i><b>28</b> Subsetting Rles</a></li>
<li class="chapter" data-level="29" data-path="subsetting-pyrles.html"><a href="subsetting-pyrles.html"><i class="fa fa-check"></i><b>29</b> Subsetting PyRles</a></li>
<li class="chapter" data-level="30" data-path="using-multiple-cores.html"><a href="using-multiple-cores.html"><i class="fa fa-check"></i><b>30</b> Using multiple cores</a></li>
<li class="chapter" data-level="31" data-path="genomicfeatures-methods-using-genomic-context.html"><a href="genomicfeatures-methods-using-genomic-context.html"><i class="fa fa-check"></i><b>31</b> GenomicFeatures: methods using genomic context</a></li>
<li class="chapter" data-level="32" data-path="fetching-gene-tracks-as-pyranges.html"><a href="fetching-gene-tracks-as-pyranges.html"><i class="fa fa-check"></i><b>32</b> Fetching gene tracks as PyRanges</a></li>
</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">PyRanges documentation</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="printing-pyranges" class="section level1">
<h1><span class="header-section-number">8</span> Printing PyRanges</h1>
<p>There are 10 helper methods for printing PyRanges. These are p/print (regular print),
mp (merged position print), sp (sorted print), msp (merged position sorted
print) and rp (raw print).</p>
<p>Merged means that the information about Chromosome, Start, End and Strand (if
applicable) are merged into one column for printing. Sorted means that the
output is sorted on position and Strand. Regular print shows the + strand first
and - strand second, so if you want to display interleaved intervals from both
strands, sorted print is the way to do it.</p>
<p>Raw print (rp) just prints the underlying dict of dataframes.</p>
<p>All print methods except rp take an n-argument to control how many entries should be displayed.</p>
<p>The pc, mpc, spc, mspc and rpc also return the pyranges so that print methods
can be used to display intermediate results in pipes.</p>
<div class="sourceCode" id="cb99"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb99-1" data-line-number="1"><span class="im">import</span> pyranges <span class="im">as</span> pr</a>
<a class="sourceLine" id="cb99-2" data-line-number="2"></a>
<a class="sourceLine" id="cb99-3" data-line-number="3">exons <span class="op">=</span> pr.data.exons()</a>
<a class="sourceLine" id="cb99-4" data-line-number="4">exons.pc(n<span class="op">=</span><span class="dv">2</span>).mpc(n<span class="op">=</span><span class="dv">8</span>).sp()</a></code></pre></div>
<pre><code>## +--------------+-----------+-----------+----------------------------------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name                                   | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)                               | (int64)   | (category)   |
## |--------------+-----------+-----------+----------------------------------------+-----------+--------------|
## | chrX         | 135721701 | 135721963 | NR_038462_exon_0_0_chrX_135721702_f    | 0         | +            |
## | ...          | ...       | ...       | ...                                    | ...       | ...          |
## | chrY         | 15467254  | 15467278  | NM_001258270_exon_13_0_chrY_15467255_r | 0         | -            |
## +--------------+-----------+-----------+----------------------------------------+-----------+--------------+
## Stranded PyRanges object has 1,000 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## +----------------------------+----------------------------------------+-----------+
## | - Position -               | Name                                   | Score     |
## | (Multiple types)           | (object)                               | (int64)   |
## |----------------------------+----------------------------------------+-----------|
## | chrX 135721701-135721963 + | NR_038462_exon_0_0_chrX_135721702_f    | 0         |
## | chrX 135574120-135574598 + | NM_001727_exon_2_0_chrX_135574121_f    | 0         |
## | chrX 47868945-47869126 +   | NM_205856_exon_4_0_chrX_47868946_f     | 0         |
## | chrX 77294333-77294480 +   | NM_000052_exon_17_0_chrX_77294334_f    | 0         |
## | ...                        | ...                                    | ...       |
## | chrY 15409586-15409728 -   | NR_047633_exon_3_0_chrY_15409587_r     | 0         |
## | chrY 15478146-15478273 -   | NR_047634_exon_18_0_chrY_15478147_r    | 0         |
## | chrY 15360258-15361762 -   | NR_047601_exon_0_0_chrY_15360259_r     | 0         |
## | chrY 15467254-15467278 -   | NM_001258270_exon_13_0_chrY_15467255_r | 0         |
## +----------------------------+----------------------------------------+-----------+
## Stranded PyRanges object has 1,000 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome and Strand.
## +--------------+-----------+-----------+---------------------------------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name                                  | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)                              | (int64)   | (category)   |
## |--------------+-----------+-----------+---------------------------------------+-----------+--------------|
## | chrX         | 585078    | 585337    | NM_000451_exon_0_0_chrX_585079_f      | 0         | +            |
## | chrX         | 1393647   | 1393735   | NM_172249_exon_1_0_chrX_1393648_f     | 0         | +            |
## | chrX         | 1393647   | 1393735   | NM_006140_exon_1_0_chrX_1393648_f     | 0         | +            |
## | chrX         | 1404670   | 1404813   | NM_172245_exon_3_0_chrX_1404671_f     | 0         | +            |
## | chrX         | 1407411   | 1407535   | NM_172249_exon_4_0_chrX_1407412_f     | 0         | +            |
## | chrX         | 1407651   | 1407781   | NM_001161532_exon_3_0_chrX_1407652_f  | 0         | +            |
## | chrX         | 1414319   | 1414349   | NM_172245_exon_8_0_chrX_1414320_f     | 0         | +            |
## | chrX         | 1419383   | 1419519   | NM_001161531_exon_9_0_chrX_1419384_f  | 0         | +            |
## | chrX         | 1424338   | 1424420   | NM_006140_exon_11_0_chrX_1424339_f    | 0         | +            |
## | chrX         | 1424338   | 1424420   | NM_001161530_exon_10_0_chrX_1424339_f | 0         | +            |
## | chrX         | 1475113   | 1475229   | NM_001267713_exon_4_0_chrX_1475114_f  | 0         | +            |
## | chrX         | 1510791   | 1511039   | NM_001636_exon_3_0_chrX_1510792_r     | 0         | -            |
## | chrX         | 1553914   | 1553976   | NM_004192_exon_8_0_chrX_1553915_r     | 0         | -            |
## | chrX         | 1554586   | 1554651   | NM_001173473_exon_9_0_chrX_1554587_r  | 0         | -            |
## | chrX         | 1734025   | 1734161   | NM_004043_exon_1_0_chrX_1734026_f     | 0         | +            |
## | ...          | ...       | ...       | ...                                   | ...       | ...          |
## | chrY         | 25336491  | 25336631  | NM_004081_exon_22_0_chrY_25336492_r   | 0         | -            |
## | chrY         | 26778138  | 26778263  | NM_001002761_exon_6_0_chrY_26778139_f | 0         | +            |
## | chrY         | 26952215  | 26952307  | NM_020364_exon_16_0_chrY_26952216_r   | 0         | -            |
## | chrY         | 26979966  | 26980276  | NM_001005375_exon_0_0_chrY_26979967_f | 0         | +            |
## | chrY         | 26998146  | 26998238  | NM_020364_exon_2_0_chrY_26998147_f    | 0         | +            |
## | chrY         | 26998798  | 26998850  | NM_001005785_exon_3_0_chrY_26998799_f | 0         | +            |
## | chrY         | 26999286  | 26999350  | NM_001005375_exon_9_0_chrY_26999287_f | 0         | +            |
## | chrY         | 27003422  | 27003494  | NM_001005786_exon_7_0_chrY_27003423_f | 0         | +            |
## | chrY         | 27041834  | 27041906  | NM_020420_exon_13_0_chrY_27041835_f   | 0         | +            |
## | chrY         | 27197822  | 27197945  | NM_004678_exon_7_0_chrY_27197823_r    | 0         | -            |
## | chrY         | 27197822  | 27197945  | NM_001002760_exon_7_0_chrY_27197823_r | 0         | -            |
## | chrY         | 27606238  | 27606322  | NR_002195_exon_8_0_chrY_27606239_f    | 0         | +            |
## | chrY         | 27606238  | 27606322  | NR_001555_exon_8_0_chrY_27606239_f    | 0         | +            |
## | chrY         | 59222126  | 59222216  | NR_033715_exon_1_0_chrY_59222127_f    | 0         | +            |
## | chrY         | 59233166  | 59233257  | NR_033715_exon_4_0_chrY_59233167_f    | 0         | +            |
## +--------------+-----------+-----------+---------------------------------------+-----------+--------------+
## Stranded PyRanges object has 1,000 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome, Start, End and Strand.</code></pre>
<div class="sourceCode" id="cb101"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb101-1" data-line-number="1">exons.<span class="bu">print</span>(n<span class="op">=</span><span class="dv">1</span>,</a>
<a class="sourceLine" id="cb101-2" data-line-number="2">            merge_position<span class="op">=</span><span class="va">False</span>,</a>
<a class="sourceLine" id="cb101-3" data-line-number="3">            sort<span class="op">=</span><span class="va">True</span>,</a>
<a class="sourceLine" id="cb101-4" data-line-number="4">            formatting<span class="op">=</span>{<span class="st">&quot;Start&quot;</span>: <span class="st">&quot;{:,}&quot;</span>, <span class="st">&quot;Name&quot;</span>: <span class="st">&quot;</span><span class="sc">{:2.2}</span><span class="st">&quot;</span>},</a>
<a class="sourceLine" id="cb101-5" data-line-number="5">            chain<span class="op">=</span><span class="va">False</span>)</a></code></pre></div>
<pre><code>## +--------------+-----------+-----------+------------+-----------+--------------+
## | Chromosome   | Start     | End       | Name       | Score     | Strand       |
## | (category)   | (int32)   | (int32)   | (object)   | (int64)   | (category)   |
## |--------------+-----------+-----------+------------+-----------+--------------|
## | ...          | ...       | ...       | ...        | ...       | ...          |
## +--------------+-----------+-----------+------------+-----------+--------------+
## Stranded PyRanges object has 1,000 rows and 6 columns from 2 chromosomes.
## For printing, the PyRanges was sorted on Chromosome, Start, End and Strand.</code></pre>
<div class="sourceCode" id="cb103"><pre class="sourceCode python"><code class="sourceCode python"><a class="sourceLine" id="cb103-1" data-line-number="1">f1 <span class="op">=</span> pr.data.f1()</a>
<a class="sourceLine" id="cb103-2" data-line-number="2">f1.rp()</a></code></pre></div>
<pre><code>## {(&#39;chr1&#39;, &#39;+&#39;):   Chromosome  Start  End       Name  Score Strand
## 0       chr1      3    6  interval1      0      +
## 2       chr1      8    9  interval3      0      +, (&#39;chr1&#39;, &#39;-&#39;):   Chromosome  Start  End       Name  Score Strand
## 1       chr1      5    7  interval2      0      -}</code></pre>

</div>
            </section>

          </div>
        </div>
      </div>
<a href="piping-pyranges.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="iterating-over-pyranges.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"download": null,
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
